binaries/daemon/src/spawn.rs

```diff
@@ -149,21 +162,82 @@ pub async fn spawn_node(
                 }
             }
 
-            command.spawn().wrap_err(format!(
-                "failed to run runtime {}/{}",
-                runtime_config.node.dataflow_id, runtime_config.node.node_id
-            ))?
+            command
+                .stdin(Stdio::null())
+                .stdout(Stdio::piped())
+                .stderr(Stdio::piped())
+                .spawn()
+                .wrap_err(format!(
+                    "failed to run runtime {}/{}",
+                    runtime_config.node.dataflow_id, runtime_config.node.node_id
+                ))?
         }
     };
 
+    let log_dir = temp_dir();
+
+    let (tx, mut rx) = mpsc::channel(10);
+    let mut file = File::create(
+        &log_dir.join(PathBuf::from(format!("{dataflow_id}-{node_id}.txt")).with_extension("txt")),
+    )
+    .await
+    .expect("Failed to create log file");
+    let mut stdout_lines =
+        (tokio::io::BufReader::new(child.stdout.take().expect("failed to take stdout"))).lines();
+
+    let stdout_tx = tx.clone();
+
+    // Stdout listener stream
+    tokio::spawn(async move {
+        while let Ok(Some(line)) = stdout_lines.next_line().await {
+            let sent = stdout_tx.send(Some(line)).await;
+            if sent.is_err() {
+                break;
+            }
+        }
+    });
+
+    let mut stderr_lines =
+        (tokio::io::BufReader::new(child.stderr.take().expect("failed to take stderr"))).lines();
+
+    // Stderr listener stream
+    let stderr_tx = tx.clone();
+    tokio::spawn(async move {
+        while let Ok(Some(line)) = stderr_lines.next_line().await {
+            let sent = stderr_tx.send(Some(line)).await;
+            if sent.is_err() {
+                break;
+            }
+        }
+    });
+
+    let exit_status_tx = tx.clone();
     tokio::spawn(async move {
         let exit_status = NodeExitStatus::from(child.wait().await);
         let event = DoraEvent::SpawnedNodeResult {
             dataflow_id,
             node_id,
             exit_status,
         };
+
+        exit_status_tx.send(None).await.unwrap();
         let _ = daemon_tx.send(event.into()).await;
     });
+
+    // Log to file stream.
+    tokio::spawn(async move {
+        while let Some(Some(line)) = rx.recv().await {
+            file.write_all(line.as_bytes())
+                .await
+                .expect("Could not log stdout/stderr to file");
+            file.write_all(b"\n")
+                .await
+                .expect("Could not add newline to log file.");
+            debug!("{dataflow_id}/{} logged {line}", node.id.clone());
+
+            // Make sure that all data has been synced to disk.
+            file.sync_all().await.unwrap();
```

Comment:
We occasionally run into these panics on the CI, for example in https://github.com/dora-rs/dora/actions/runs/4805449241/jobs/8551873559#step:11:619 . This is something that we should fix before merging.

Reply:
Indeed. 

So the issue was that in previous iteration of logging, the logging thread was not waited for and the daemon would exit before finishing to log all lines causing a panic. 

You can have a more detailed error here: https://github.com/dora-rs/dora/actions/runs/4862287019/jobs/8668444816

The error is exactly the same as here: https://github.com/tokio-rs/tokio/issues/2247

I have managed to fix that issue by creating a oneshot channel that would wait for the logging thread to finish before marking the node being logged as finished.